be/src/exec/pipeline/stream_pipeline_driver.cpp

```diff
@@ -0,0 +1,332 @@
+#include "exec/pipeline/pipeline_fwd.h"
+#include "exec/pipeline/runtime_filter_types.h"
+#include "exec/pipeline/scan/morsel.h"
+#include "exec/workgroup/work_group.h"
+#include "exec/workgroup/work_group_fwd.h"
+
+namespace starrocks::pipeline {
+
+StatusOr<DriverState> StreamPipelineDriver::process(RuntimeState* runtime_state, int worker_id) {
+    COUNTER_UPDATE(_schedule_counter, 1);
+    SCOPED_TIMER(_active_timer);
+    QUERY_TRACE_SCOPED("process", _driver_name);
+    set_driver_state(DriverState::RUNNING);
+    size_t total_chunks_moved = 0;
+    size_t total_rows_moved = 0;
+    int64_t time_spent = 0;
+    Status return_status = Status::OK();
+    StatusOr<DriverState> resturn_state = Status::OK();
+    DeferOp defer([&]() {
+        if (return_status.ok()) {
+            _update_statistics(total_chunks_moved, total_rows_moved, time_spent);
+        }
+    });
+
+    bool should_yield = false;
+    size_t num_chunks_moved = 0;
+    size_t num_operators = _operators.size();
+    size_t new_first_epoch_unfinished = _first_epoch_unfinished;
+    while (true) {
+        RETURN_IF_LIMIT_EXCEEDED(runtime_state, "Pipeline");
+
+        // `finished` is only used to stop the MV maintenance and should be
+        // runned without blocking because there is no data processing.
+        if (UNLIKELY(_operators[_first_unfinished]->is_finished())) {
+            return _handle_finish_operators(runtime_state);
+        }
+
+        for (size_t i = _first_epoch_unfinished; i < num_operators - 1; ++i) {
+            {
+                SCOPED_RAW_TIMER(&time_spent);
+                auto& curr_op = _operators[i];
+                auto& next_op = _operators[i + 1];
+
+                // Check curr_op epoch_finished firstly
+                if (curr_op->is_epoch_finished()) {
+                    if (i == 0) {
+                        // For source operators
+                        RETURN_IF_ERROR(return_status = _mark_operator_epoch_finishing(curr_op, runtime_state));
+                    }
+                    RETURN_IF_ERROR(return_status = _mark_operator_epoch_finishing(next_op, runtime_state));
+                    new_first_epoch_unfinished = i + 1;
+                    continue;
+                }
+
+                // try successive operator pairs
+                if (!curr_op->has_output() || !next_op->need_input()) {
+                    continue;
+                }
+
+                if (_check_fragment_is_canceled(runtime_state)) {
+                    return _state;
+                }
+
+                // pull chunk from current operator and push the chunk onto next
+                // operator
+                StatusOr<ChunkPtr> maybe_chunk;
+                {
+                    SCOPED_TIMER(curr_op->_pull_timer);
+                    QUERY_TRACE_SCOPED(curr_op->get_name(), "pull_chunk");
+                    maybe_chunk = curr_op->pull_chunk(runtime_state);
+                }
+                return_status = maybe_chunk.status();
```

Comment:
so much code are duplicated with `PipelineDriver`, we need to consider some refactor